### PR TITLE
Optimize perf of softmax_with_cross_entropy

### PR DESCRIPTION
**Pull request by ZzSean**
### PR types

Performance optimization
### PR changes

OPs
### Describe

Optimize perf of softmax_with_cross_entropy

Replace cudnn with CUDA kernel, and make the process of cross entrop into softmax 

| case                                                         | pytorch | paddle 优化前 | diff          | paddle 优化后 | diff         | 加速比 |
| ------------------------------------------------------------ | ------- | ------------- | ------------- | ------------- | ------------ | ------ |
| label(Variable)-dtype:int64,shape:[2,1024,1024,1]       logits(Variable)-dtype:float32,shape:[2,1024,1024,19]      axis(int):3 ignore_index(int):255 soft_label(bool):False | 0.71130 | 0.71989       | 打平 (1.21%)  | 0.71510       | 打平 (0.53%) | 1.01   |
| label(Variable)-dtype:int64,shape:[8,1024,1]     logits(Variable)-dtype:float32,shape:[8,1024,50257]      axis(int):2 ignore_index(int):-100 soft_label(bool):False | 8.20172 | 9.96242       | 差于 (21.47%) | 8.32200       | 打平 (1.47%) | 1.20   |
| label(Variable)-dtype:int64,shape:[8,1024,1]       logits(Variable)-dtype:float32,shape:[8,1024,50304]      axis(int):2 ignore_index(int):-100 soft_label(bool):False | 8.01143 | 9.27125       | 差于 (15.73%) | 8.07147       | 打平 (0.75%) | 1.15   |
| label(Variable)-dtype:int64,shape:[8,1024,1]       logits(Variable)-dtype:float16,shape:[8,1024,50304]      axis(int):2 ignore_index(int):-100 soft_label(bool):False | 3.99666 | 9.66784       | 差于 (1.42x)  | 4.20327       | 差于 (5.17%) | 2.30   |

### CODE REVIEW

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 1]
```diff
@@ -419,10 +433,272 @@ void SwitchWarpSoftmaxForward(T* loss, T* softmax, const T* src,
...
+                                            const int ignore_index) {
+  int loss_id = vec_size * tid + offset;
+  if (IgnoreIndex) {
+    if (label_value == loss_id) {
+      if (label_value == ignore_index) {
+        loss[label_id] = static_cast<T>(0.0f);
+      } else {
+        loss[label_id] = loss_value;
+      }
+    }
```
Comment (xingfeng01) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:451]: else 的情况会有问题吗？好像有些情况没有给 loss 赋值

**Review by yghstill — APPROVED** [review]

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 2]
```diff
@@ -419,10 +433,272 @@ void SwitchWarpSoftmaxForward(T* loss, T* softmax, const T* src,
...
+
+  // scalar part
+  tid = size - remain + threadIdx.x;
+  for (; tid < size; tid += blockDim.x) {
+    val = reducer(val, input[tid]);
+  }
+  return val;
+}
+
+template <typename T, typename AccT, typename LabelT, int VecSize,
```
Comment (xingfeng01) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:499]: 建议给函数加点注释，说明一下函数功能

Comment (ZzSean) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:499]: 好的，下个PR会加一些注释。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 1 above)
Comment (ZzSean) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu:451]: label_value的有效范围是[0, size)，而loss_id的遍历范围也是[0, size)，因此不会出现loss不被赋值的情况

**Review by xingfeng01 — APPROVED** [review]
